### PR TITLE
cluster agent: create a cluster agent client 

### PR DESCRIPTION
**Pull request by JulienBalestra**
### What does this PR do?

Create a cluster agent util to query the Datadog Cluster Agent (DCA) API.

new configurations entries:
* cluster_agent_auth_token
* cluster_agent_url
* cluster_agent_kubernetes_service_name

The client will use **https** only to query the DCA.
As first implementation, the DCA will expose exclusively an **https** self-signed endpoint.

##### Current usage
The DCA API will expose a route where you can get the Kubernetes service names associated to a given Pod.

The public method `GetKubernetesServiceNames` give to the client the capabilities to call this API.

### CODE REVIEW

File: pkg/api/util/clusteragent.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	const configDcaURL = "cluster_agent_url"
+	const configDcaSvcName = "cluster_agent_kubernetes_service_name"
+
+	dcaURL := config.Datadog.GetString(configDcaURL)
+	if dcaURL != "" {
+		if strings.HasPrefix(dcaURL, "http://") {
+			return "", fmt.Errorf("cannot get cluster agent endpoint, not a https scheme: %s", dcaURL)
+		}
+		if strings.Contains(dcaURL, "://") == false {
+			log.Tracef("adding https scheme to %s: https://%s", dcaURL, dcaURL)
```
Comment (CharlyF) [inline pkg/api/util/clusteragent.go]: I think `adding https scheme to %s` suffices

File: pkg/api/util/clusteragent.go [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+	// Construct the URL with the Kubernetes service environment variables
+	// *_SERVICE_HOST and *_SERVICE_PORT
+	dcaSvc := config.Datadog.GetString(configDcaSvcName)
+	if dcaSvc == "" {
+		return "", fmt.Errorf("cannot get a cluster agent endpoint, both %q and %q are empty", configDcaURL, configDcaSvcName)
+	}
+
+	dcaSvc = strings.ToUpper(dcaSvc)
+
+	// host
```
Comment (CharlyF) [inline pkg/api/util/clusteragent.go]: I think extracting from the env and properly dealing with the "" is great (also for the //port)

File: pkg/api/util/clusteragent.go [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+	// TODO remove insecure
+	c.clusterAgentAPIClient = GetClient(false)
+	c.clusterAgentAPIClient.Timeout = 2 * time.Second
+
+	return nil
+}
+
+// getClusterAgentEndpoint provides a validated https endpoint from:
+// 1. configuration key "cluster_agent_url" like https
```
Comment (CharlyF) [inline pkg/api/util/clusteragent.go]: we should precise that this value has to be set in the datadog.yaml

File: pkg/api/util/clusteragent.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+// Kubernetes services.
+func (c *ClusterAgentUtil) GetKubernetesServiceNames(nodeName, podName string) ([]string, error) {
+	const dcaMetadataPath = "api/v1/metadata"
+	var serviceNames serviceNames
+	var err error
+
+	req := &http.Request{
+		Header: *c.clusterAgentAPIRequestHeaders,
+	}
+	// https://host:port /api/v1/metadata/ {nodeName}/ {pod-[0-9a-z]{5}}
```
Comment (CharlyF) [inline pkg/api/util/clusteragent.go]: Not sure about this regex. If I'm not misreading this: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/ it seems like the podname can be much longer

File: pkg/api/util/clusteragent.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+	if resp.StatusCode != http.StatusOK {
+		return serviceNames, fmt.Errorf("unexpected status code from cluster agent: %d", resp.StatusCode)
+	}
+
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return serviceNames, err
+	}
+	err = json.Unmarshal(b, &serviceNames)
```
Comment (CharlyF) [inline pkg/api/util/clusteragent.go]: Perfect, as we are marshalling upstream .

**Review by CharlyF — COMMENTED** [review]
This is great 🌮 . On a side note, I think we need the DNA to check regularly if the DCA endpoint has been referenced or is detected. Just so we don't need to restart should the DCA be started later on.

File: pkg/api/util/clusteragent.go (hunk 1 above)
Comment (JulienBalestra) [reply pkg/api/util/clusteragent.go]: If the user specify `tcp://` we insert involuntary `https://tcp://...`
I'll see how to improve this part

Comment (JulienBalestra) [reply pkg/api/util/clusteragent.go]: See [tcp](https://github.com/DataDog/datadog-agent/pull/1127/files#diff-caa70174faffdcf870b85dd14e28883dR153) error handling

File: pkg/api/util/clusteragent.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2018 Datadog, Inc.
+
+package util
```
Comment (CharlyF) [inline pkg/api/util/clusteragent.go]: Could we rename this to clusteragent ? To be consistent with the other utils in the pkg ?

File: pkg/api/util/clusteragent.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+	log "github.com/cihub/seelog"
+
+	"github.com/DataDog/datadog-agent/pkg/config"
```
Comment (mfpierre) [inline pkg/api/util/clusteragent.go]: should be grouped with above dependencies

File: pkg/api/util/clusteragent.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	if err != nil {
+		return err
+	}
+
+	authToken := config.Datadog.GetString("cluster_agent_auth_token")
+	if authToken == "" {
+		return fmt.Errorf("empty cluster_agent_auth_token")
+	}
+	if len(authToken) < clusterAgentAuthTokenMinLen {
+		return fmt.Errorf("need at least a length of %d for cluster_agent_auth_token: %d", clusterAgentAuthTokenMinLen, len(authToken))
```
Comment (mfpierre) [inline pkg/api/util/clusteragent.go]: `specified token is %d` would be more explicit

File: pkg/api/util/clusteragent.go (hunk 6 above)
Comment (JulienBalestra) [reply pkg/api/util/clusteragent.go]: I agree with that change, @DataDog/kenafeh any thoughts ?

Comment (xvello) [reply pkg/api/util/clusteragent.go]: indeed. ideally, it should have it's build tags as it's not needed for the puppy agent (for the moment?)

File: pkg/config/config_template.yaml [hunk 9]
```diff
@@ -306,6 +306,30 @@ metadata_collectors:
 #
 {{ end -}}
 
+# Datadog Cluster Agent - DCA
+#
+# The Authentication Token used to query the Datadog cluster agent from the node agent.
+# This token is provided in the header "Authorization" of each request, its length be must
```
Comment (olivielpeau) [inline pkg/config/config_template.yaml]: `its length be must` -> `its length must be`

also, not sure we should mention that it's in the "Authorization" header, to me it's more of an implementation detail/should be documented somewhere else

File: pkg/config/config_template.yaml [hunk 10]
```diff
@@ -306,6 +306,30 @@ metadata_collectors:
...
+# The only supported scheme to reach the Datadog cluster agent is https, example: "https://host:port"
+# cluster_agent_url: ""
+
+# If the node-agent runs inside a Kubernetes Pod, you can use a Kubernetes Service to discover the
+# cluster agent.
+# To select this configuration, you need to leave the cluster_agent_url unset ("") and provide the
+# name of the Kubernetes service.
+# cluster_agent_kubernetes_service_name: "dca"
+#
+# The cluster_agent_url is evaluated before cluster_agent_kubernetes_service_name.
```
Comment (olivielpeau) [inline pkg/config/config_template.yaml]: this also sounds like an implementation detail (the sentence above, `To select this configuration [...]`, is a bit redundant with this one)

File: pkg/config/config_template.yaml [hunk 11]
```diff
@@ -306,6 +306,30 @@ metadata_collectors:
 #
 {{ end -}}
 
+# Datadog Cluster Agent - DCA
```
Comment (olivielpeau) [inline pkg/config/config_template.yaml]: This block of documentation should live under a "context" (see https://github.com/DataDog/datadog-agent/blob/dd5c72fc063c1b09b60790e369ff03435bba73f0/pkg/config/render_config.go#L19-L34).

if it should be common to all builds (agent, dogstatsd and dca), please use `Common`, or feel free to add a more appropriate "context" name :)

Comment (olivielpeau) [reply pkg/config/config_template.yaml]: Also, can you add a quick explanation on when a user should care at all about the `Datadog Cluster Agent` settings?

File: pkg/config/config.go [hunk 12]
```diff
@@ -145,6 +145,11 @@ func init() {
 	// Kube ApiServer
 	Datadog.SetDefault("kubernetes_kubeconfig_path", "")
 
+	// Datadog cluster agent
+	Datadog.SetDefault("cluster_agent_auth_token", "")
+	Datadog.SetDefault("cluster_agent_url", "")
+	Datadog.SetDefault("cluster_agent_kubernetes_service_name", "dca")
```
Comment (olivielpeau) [inline pkg/config/config.go]: maybe `cluster_agent` options are a good candidate to be nested options, e.g.:
```yaml
cluster_agent:
  auth_token: aaaaaaaaaa
  url: https://foo:12345
```

and the related values can be set/accessed with, for example: `Datadog.SetDefault("cluster_agent.auth_token", "")`, `Datadog.Get("cluster_agent.auth_token", "")`

Comment (olivielpeau) [inline pkg/config/config.go]: also, didn't we want to bind `cluster_agent_auth_token` to the related env var?

File: pkg/util/clusteragent/clusteragent.go [hunk 13]
```diff
@@ -0,0 +1,223 @@
...
+	dcaURL = fmt.Sprintf("https://%s:%s", dcaSvcHost, dcaSvcPort)
+	u, err := url.Parse(dcaURL)
+	if err != nil {
+		return "", err
+	}
+
+	return u.String(), nil
+}
+
+// GetKubernetesServiceNames query the datadog cluster agent to get nodeName/podName registered
```
Comment (olivielpeau) [inline pkg/util/clusteragent/clusteragent.go]: `queries`

File: pkg/util/clusteragent/clusteragent.go [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+	clusterAgentAuthTokenMinLen   = 32
+	clusterAgentAuthTokenFilename = "dca_auth_token"
+)
+
+var globalClusterAgentUtil *DCAUtil
+
+type serviceNames []string
+
+// DCAUtil is required to query the API of Datadog cluster agent
+type DCAUtil struct {
```
Comment (olivielpeau) [inline pkg/util/clusteragent/clusteragent.go]: any reason to call this `Util` instead of `Client`?

**Review by olivielpeau — COMMENTED** [review]
A few comments here and there, overall this LGTM 👍

File: pkg/config/config.go (hunk 12 above)
Comment (JulienBalestra) [reply pkg/config/config.go]: Indeed, that's better ! :+1:

Comment (JulienBalestra) [reply pkg/config/config.go]: Yes, do we add the others ?

Comment (olivielpeau) [reply pkg/config/config.go]: Don't know the use cases of having the other ones as env vars, up to you :)

File: pkg/util/clusteragent/clusteragent.go (hunk 14 above)
Comment (JulienBalestra) [reply pkg/util/clusteragent/clusteragent.go]: Indeed, I picked `Client` in the first place.
I fallback to `Util` to be consistent with the others utils...

**Review by olivielpeau — APPROVED** [review]
🚀

File: pkg/config/config.go (hunk 12 above)
Comment (JulienBalestra) [reply pkg/config/config.go]: We'll see when using it...

**Pull request merged**